### DISCUSSION TITLE
Prometheus metrics

### DISCUSSION THREAD

**Discussion by jinnatar** [PremoWeb/chadburn#4] [Ideas]
At current it's a bit annoying to monitor chadburn to ensure that everything is working as intended. Even a basic health check is a bit hard to do since the service doesn't expose any TCP port (since there's no need), so there's no port to check for a response.

A basic Prometheus metrics endpoint would solve for the basic "alive" check by providing an `up` metric and could be expanded to contain more useful metrics such as:
- successful executions counter
- failed executions counter
- current registered services count
- Execution time metrics

**Comment by maietta** [comment]
Sorry for the late response. I'm switching email systems as I haven't been getting all my Github notifications.

I like your suggestion. PremoWeb prioritizes man hours on open source projects based on interest and incentives like paid sponsorships through Github. If this is something you need much sooner, a contribution can push this feature request higher up on the list and get it done quicker but is not required.

I will schedule some time in this next week to get started on this. Thank you for the feature suggestion.

**Reply by jinnatar** [reply]
That makes sense. I've never written a Prometheus endpoint myself but want to learn so might try to implement myself as well. Shouldn't be all that hard since both are Go.

**Reply by maietta** [reply]
I did some studying and have a the server deployed. I might have a solution tomorrow, or at least most of it figured out.

**Reply by jinnatar** [reply]
Did you get anything semi functional written? If you push it to a feature branch I could take a look and see if I can continue from there.

**Reply by maietta** [reply]
Thanks for reminding me about it. (I forgot to finish getting this working). I will be back at my office in a couple of days where I may be able to continue working on this. For the record, while I did study Prometheus, it seemed a bit convoluted and I wasn't able to get anything to report to Prometheus as of yet from any instructions I found anywhere online.

**Reply by maietta** [reply]
Similar issue raised here: https://github.com/PremoWeb/chadburn/issues/10

**Reply by jinnatar** [reply]
Prometheus is a bit of a different cookie since it's all about metrics. For development you don't necessarily need a Prometheus server. All chadburn needs to do is use a client library to expose an endpoint that gives instant metrics at that point in time. Ideally the data for it is collected as things happen, which means the metrics endpoint just needs to read from memory. If the endpoint gives instant stats that's already valuable. Prometheus itself only adds the time axis to it by collecting and storing them over time.

**Comment by maietta** [comment]
Still in the works...

**Comment by jinnatar** [comment]
Had some idle time on my hands so made a basic implementation in #38.

**Reply by maietta** [reply]
@Artanicus, Thank you so much for your contribution. You are awesome!

After code review and local testing, everything looked good enough to go ahead and merge your changes. I'll get the readme updated shortly.

Note that I did have to remove an old version of Go 1.15.x from testing and just went ahead and only referenced the latest stable version of Go from here on. Not sure that we needed multiple versions of Go for testing anyway-- seems pointless in the case of Chadburn.
